internal/testpeertls/revocations.go

```diff
@@ -8,50 +8,88 @@ import (
-	var revocation pkix.Extension
-	revokingKey, err := pkcrypto.GeneratePrivateKey()
-	if err != nil {
-		return nil, revocation, err
+func RevokeLeaf(caKey crypto.PrivateKey, chain []*x509.Certificate) ([]*x509.Certificate, pkix.Extension, error) {
+	if len(chain) < 2 {
+		return nil, pkix.Extension{}, extensions.Error.New("revoking leaf implies a CA exists; chain too short")
+	}
+	ca := &identity.FullCertificateAuthority{
+		Key:       caKey,
+		Cert:      chain[peertls.CAIndex],
+		RestChain: chain[:peertls.CAIndex+1],
 	}
 
-	revokingTemplate, err := peertls.LeafTemplate()
+	var err error
+	ca.ID, err = identity.NodeIDFromKey(ca.Cert.PublicKey)
 	if err != nil {
-		return nil, revocation, err
+		return nil, pkix.Extension{}, err
 	}
 
-	revokingCert, err := peertls.NewCert(revokingKey, keys[0], revokingTemplate, chain[peertls.CAIndex])
-	if err != nil {
-		return nil, revocation, err
+	ident := &identity.FullIdentity{
+		Leaf:      chain[peertls.LeafIndex],
+		CA:        ca.Cert,
+		ID:        ca.ID,
+		RestChain: ca.RestChain,
 	}
 
-	err = extensions.AddRevocationExt(keys[0], chain[peertls.LeafIndex], revokingCert)
-	if err != nil {
-		return nil, revocation, err
+	manageableIdent := identity.NewManageableFullIdentity(ident, ca)
+	if err := manageableIdent.Revoke(); err != nil {
+		return nil, pkix.Extension{}, err
+	}
+
+	revokingCert := manageableIdent.Leaf
+	var revocationExt *pkix.Extension
+	for _, ext := range revokingCert.Extensions {
+		if extensions.RevocationExtID.Equal(ext.Id) {
+			revocationExt = &ext
+			break
+		}
+	}
+	if revocationExt == nil {
+		return nil, pkix.Extension{}, extensions.ErrRevocation.New("no revocation extension found")
 	}
 
-	revocation = revokingCert.ExtraExtensions[0]
-	return append([]*x509.Certificate{revokingCert}, chain[peertls.CAIndex:]...), revocation, nil
+	return append([]*x509.Certificate{revokingCert}, chain[peertls.CAIndex:]...), *revocationExt, nil
 }
 
 // RevokeCA revokes the CA certificate in the passed chain and adds a revocation
 // extension to that certificate, recording this action.
-func RevokeCA(keys []crypto.PrivateKey, chain []*x509.Certificate) ([]*x509.Certificate, pkix.Extension, error) {
+func RevokeCA(caKey crypto.PrivateKey, chain []*x509.Certificate) ([]*x509.Certificate, pkix.Extension, error) {
 	caCert := chain[peertls.CAIndex]
-	err := extensions.AddRevocationExt(keys[0], caCert, caCert)
+	nodeID, err := identity.NodeIDFromKey(caCert.PublicKey)
 	if err != nil {
 		return nil, pkix.Extension{}, err
 	}
 
-	return append([]*x509.Certificate{caCert}, chain[peertls.CAIndex:]...), caCert.ExtraExtensions[0], nil
+	ca := &identity.FullCertificateAuthority{
+		ID:        nodeID,
+		Cert:      caCert,
+		Key:       caKey,
+		RestChain: chain[peertls.CAIndex+1:],
+	}
+
+	if err = ca.Revoke(); err != nil {
+		return nil, pkix.Extension{}, err
+	}
+
+	var revocationExt *pkix.Extension
+	for _, ext := range ca.Cert.Extensions {
```

Comment:
maybe pull this out into a sep function call?

Reply:
good call - there's actually a function that does most of this already.